File: src/services/imageProcessor.js
import fetch from 'node-fetch';
import sharp from 'sharp';

export class ImageProcessor {
  async processJob(jobId, visits, jobService) {
    const errors = [];

    try {
      for (const visit of visits) {
        for (const imageUrl of visit.image_url) {
          try {
            await this.processImage(imageUrl);
          } catch (error) {
            errors.push({
              store_id: visit.store_id,
              error: error.message
            });
            jobService.updateJobStatus(jobId, 'failed', errors);
            return;
          }
        }
      }

      jobService.updateJobStatus(jobId, 'completed');
    } catch (error) {
      jobService.updateJobStatus(jobId, 'failed', [{
        error: 'Internal server error'
      }]);
    }
  }

  async processImage(url) {
    // Download image
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error('Failed to download image');
    }

    const buffer = await response.buffer();
    
    // Get image dimensions using sharp
    const metadata = await sharp(buffer).metadata();
    const perimeter = 2 * (metadata.width + metadata.height);

    // Simulate GPU processing time
    await new Promise(resolve => 
      setTimeout(resolve, 100 + Math.random() * 300)
    );

    return perimeter;
  }
}

File: service/image_processor.go
package service

import (
	"image"
	_ "image/jpeg"
	_ "image/png"
	"image-processor/models"
	"math/rand"
	"net/http"
	"time"
)

type ImageProcessor struct {
	jobService *JobService
}

func NewImageProcessor() *ImageProcessor {
	return &ImageProcessor{}
}

func (p *ImageProcessor) ProcessJob(jobID string, visits []models.Visit) {
	var errors []models.JobError

	for _, visit := range visits {
		// Process each image URL
		for _, imageURL := range visit.ImageURLs {
			if err := p.processImage(imageURL); err != nil {
				errors = append(errors, models.JobError{
					StoreID: visit.StoreID,
					Error:   err.Error(),
				})
				p.jobService.UpdateJobStatus(jobID, models.StatusFailed, errors)
				return
			}
		}
	}

	p.jobService.UpdateJobStatus(jobID, models.StatusCompleted, nil)
}

func (p *ImageProcessor) processImage(url string) error {
	// Download image
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Decode image
	img, _, err := image.Decode(resp.Body)
	if err != nil {
		return err
	}

	// Calculate perimeter
	bounds := img.Bounds()
	perimeter := 2 * (bounds.Dx() + bounds.Dy())

	// Random sleep to simulate GPU processing
	sleepTime := 100 + rand.Intn(301) // 100-400ms
	time.Sleep(time.Duration(sleepTime) * time.Millisecond)

	return nil
}